pkg/jobstore/boltdb/store.go

```diff
@@ -690,11 +699,94 @@ func (b *BoltJobStore) CreateJob(ctx context.Context, job models.Job, event mode
 	if err != nil {
 		return err
 	}
-	return b.database.Update(func(tx *bolt.Tx) (err error) {
+	return b.update(ctx, func(tx *bolt.Tx) (err error) {
 		return b.createJob(tx, job, event)
 	})
 }
 
+// update is a helper function that will update the job in the store
+// it accepts a context, an update function and creates a new transaction to
+// perform the update if no transaction is provided in the context
+func (b *BoltJobStore) update(ctx context.Context, update func(tx *bolt.Tx) error) error {
+	var err error
+	var tx *bolt.Tx
+	var managed bool
+
+	// if ctx has a transaction value, then we can use that transaction
+	// and mark it as managed so that we don't try to commit it or rollback in this function
+	var txFound bool
+	tx, txFound = txFromContext(ctx)
+	if txFound {
+		if !tx.Writable() {
+			return fmt.Errorf("readonly transaction provided in context for update operation")
+		}
+		managed = true
+	} else {
+		tx, err = b.database.Begin(true)
+		if err != nil {
+			return err
+		}
+	}
+
+	// always rollback the transaction if we did create it and we're returning an error
+	defer func() {
+		if !managed && err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
+	err = update(tx)
+	if err != nil {
+		return err
+	}
+
+	// if we created the transaction, then we need to commit it
+	if !managed {
+		err = tx.Commit()
+	}
+	return err
+}
+
+// view is a helper function that will perform a read-only operation on the store
+// it accepts a context, a view function and creates a new transaction to
+// perform the view if no transaction is provided in the context
+func (b *BoltJobStore) view(ctx context.Context, view func(tx *bolt.Tx) error) error {
+	var err error
+	var tx *bolt.Tx
+	var managed bool
+
+	// if ctx has a transaction value, then we can use that transaction
+	// and mark it as managed so that we don't try to commit it or rollback in this function
+	var txFound bool
+	tx, txFound = txFromContext(ctx)
+	if txFound {
+		managed = true
+	} else {
+		tx, err = b.database.Begin(false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// always rollback the transaction if we did create it and we're returning an error
+	defer func() {
+		if !managed && err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
+	err = view(tx)
+	if err != nil {
+		return err
+	}
+
+	// if we created the transaction, then we need to commit it
+	if !managed {
+		err = tx.Rollback()
+	}
```

Comment:
Is this comment wrong, or is the code incorrect?

Reply:
The comment is wrong as we don't commit readonly transactions, just rollback them to release the lock. Fixed in https://github.com/bacalhau-project/bacalhau/pull/3996/commits/f2180af41a362b9bcc94050facc9838a48f25651